src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs

```diff
@@ -80,15 +80,12 @@ internal sealed class State
 
             public void RemoveStream(MsQuicStream? stream)
             {
-                bool releaseHandles;
-                lock (this)
-                {
-                    StreamCount--;
-                    Debug.Assert(StreamCount >= 0);
-                    releaseHandles = _closing && StreamCount == 0;
-                }
+                int count = Interlocked.Decrement(ref StreamCount);
```

Comment:
Why do we need to change the logic in RemoveStream and TryAddStream to not take the lock? We aren't calling in to MsQuic here, so why does it matter if we take the lock?

Reply:
We don't have to solve the deadlock. But since I was looking at the locking I wanted to decrease the contention from the NativeCallbackHandler.

Reply:
There are a bunch of places we take the lock in the callback, right? I'm not sure changing this one really matters.

Reply:
yes. However this one was prevalent and somewhat not obvious coming to Connection from Stream operations. 
The new part looks simpler and safer. So I'm waiting for @stephentoub to explain more on this suggestion. 
I'm also fine if we dump this part for now.

Reply:
The changes here in RemoveStream/TryAddStream are much more involved/complicated than the basic fix to WaitForAvailableBidirectionalStreamsAsync. Let's get the latter in and we can consider the former in another PR if we think it's worthwhile.

Reply:
> to explain more on this suggestion

I meant that rather than doing:
```C#
Interlocked.Increment(ref count);
```
which could increment even after closure, you do use CompareExchange in a loop to avoid incrementing if the count is negative, e.g.
```C#
while (true)
{
    int old = Volatile.Read(ref count);
    if (old < 0) break;
    if (Interlocked.CompareExchange(ref count, old + 1, old) == old) break;
}
```

That said, I agree with Geoff to make the minimal fix necessary, changing as little synchronization as is necessary, and then revisit the whole mechanism subsequently.

Reply:
ok. Rolled back the Add/Remove changes. Added Assert for lock.